### DISCUSSION TITLE
NewGRF House Spec Improvements

### DISCUSSION THREAD

**Discussion by 2TallTyler** [OpenTTD/OpenTTD#12281] [Features]
There's been some discussion on Discord recently about improvements to the NewGRF house spec. Let's not lose these ideas.

(Each suggestion has its own post below, allowing replies to be made to that suggestion.)

**Comment by 2TallTyler** [comment]
## Larger multi-tile houses

Houses must have a substitute base game house ID, and multi-tile houses must use the IDs of a similarly-sized base game house. This limits available houses to 1x1, 1x2, 2x1, or 2x2 tiles.

It would be nice to allow GRF houses to be larger than 2x2 tiles.

**Reply by PeterN** [reply]
We could change the spec to allow multi-tile houses to use the same ID, but if that requires more map storage then it's actually better to just use consecutive IDs like now.

**Reply by 2TallTyler** [reply]
Consecutive IDs are fine with a higher limit, more of an issue is requiring the substitute ID to be a base game house of the same size. There are also some hardcoded functions in `town_cmd.cpp` that would need to be adapted for larger multi-tile houses.

**Comment by PeterN** [comment]
## Increase total number of house tile types.

I now have a patch for this that increases the limit to 4096, using 12 bits instead of 9 bits of map storage. More is possible I suppose...

**Reply by 2TallTyler** [reply]
Yes, I was posting my next comment about this. 🙂 

For reference, house IDs go up to 255, and (as you posted on Discord -- I haven't verified this) when multiple GRFs are used there's a total limit of 512 houses in a game. Multi-tile houses use one ID per tile.

**Reply by PeterN** [reply]
This is implemented with #12287 and #12288.

**Reply by reldred** [reply]
I can't remember how many bits we (and by we I mean, JGR told me what buttons to push) stole for JGRPP, but we're at either 11 bits or 12 bits total. Works good. The default 512 gets eaten up by the vanilla houses even if they're disabled so it's rather easy to smack into that limit.

**Comment by 2TallTyler** [comment]
## Upgrade path callback

The biggest problem with multi-tile houses is that when a house is rebuilt, it is demolished and then a substitute house is randomly chosen to take its place. When a multi-tile house is demolished, the house which takes its place is often smaller, leading to the other tiles being filled in by other houses. This space will never again have a multi-tile house, because the space is gone.

We could add a callback, called when a house is about to be demolished, that allows that house to choose its replacement. Multi-tile houses could only allow other multi-tile houses to replace them, keeping that space reserved. This also allows for other fun features like house upgrades and modernization, and whatever else GRF authors can dream up. 😄

**Reply by PeterN** [reply]
House upgrades is a compelling feature here, as otherwise the game could simply prefer (or have a higher chance) to build multi-tile buildings if one was removed.

**Reply by 2TallTyler** [reply]
I tried higher-probability multi-tile houses in ITL houses, but it built too many of them when the town was expanding. 😉 I settled on protecting multi-tile houses from demolition entirely, which works but isn't particularly elegant.

**Reply by nielsmh** [reply]
Allowing an upgrade/replacement callback to block the attempt to be replaced entirely might also have value. For example, if a building only available before 1930 (and maybe has an appropriate random seed) is attempted replaced in year 2030, i.e. has survived for a hundred years, it could block being replaced for "being historical". Of course that would not prevent players from demolishing it, which would be an entirely different concern.

**Reply by PeterN** [reply]
Why "of course"?

**Reply by zephyris** [reply]
... or allow removal of town buildings to build a multi-tile building?

**Reply by ldpl** [reply]
Town growth works on house count so allowing one building to replace many would affect growth poorly, effectively stopping it much sooner that current soft limit of around 25k houses. Though there is a possible solution of counting those removed houses and rebuilding them "for free", i.e. in addition to the regular growth loop.

**Reply by 2TallTyler** [reply]
> Allowing an upgrade/replacement callback to block the attempt to be replaced entirely might also have value.

We already have a callback for that, called `protection` in NML. ITL Houses uses it for historical districts already, often allowing a historical town center to be separate from the "new town" with skyscrapers that spawns elsewhere in a large city.

Unrelated but also interesting, there's also a `destruction` callback which attempts to destroy the house early.

**Reply by Ragin1924** [reply]
You could do some fun things with this. You could also allow for properties to be subdivided like you see irl. E.g. a 4x4 farm style house on the edge of town has a chance to become 4 1x1 houses when it becomes part of an inner town zone.

**Reply by reldred** [reply]
Oh I'd love this. This would be fantastic for a few sets I'm building - for skyscrapers, I can make sure they get replaced with similar buildings, for piddly little things I can work on a level progression system like Simcity and Cities Skylines.

**Reply by PikkaBird** [reply]
I wonder if #7604 didn't inadvertently make the placing of larger houses more difficult, if it means larger buildings cannot directly replace other larger buildings except on the southern side of the town.

**Comment by KeepinItRail** [comment]
### Callback for type of service
Currently there is a callback which tells you if e.g. PASS have been delivered in current or previous month. Could there be a way to determine if this is road, rail etc service?

Thinking being a bus service is not good enough to allow skyscrapers, but a rail station would be.

**Reply by zephyris** [reply]
An alternative/simpler but related variables could be number of stations in the town, maybe broken down by station type, or number of company-owned tiles in the town. Subject to performance concerns of course.

**Reply by KeepinItRail** [reply]
This would work, especially for smaller towns, at the 'town level'.

I was thinking more along the lines of `cargo_accepted_nearby_last_month(PASS)` only in addition to cargo type, adding the transportation type, which would apply more at the station catchment level. 

What might work instead is what you suggested, `number_passengers_transported` or combining with above for something like `number_passengers_transported_nearby_last_month` so you could 'localise' it to catchment area.

**Comment by KeepinItRail** [comment]
### Population of nearby cities
Players often grow cities into each other, making a larger mega city. However the spec just allows you to get the population of city of the relevant tile. Is there a way to get the population of the larger mega city? E.g. return the population of all cities within n tiles.

**Comment by zephyris** [comment]
## Height and bridging
It would be great if bridges could span over houses, so long as the house is short enough and the bridge is high enough. This is similar to behaviour of objects since OpenTTD 1.1, where a combination of an object flag and a height define if an object can have a bridge above it:

A `building_flags` (property `19`) value allowing bridges over the house.

... paired with...

A `height` property for town houses (like property `16` for objects) defining the height of the house in heightlevels (8 pixels). This determines how high a bridge must be if it is allowed by the bit in property `19`.

**Reply by zephyris** [reply]
Bridging should either always be disallowed for larger than 1x1 houses, to avoid supports in the middle of buildings. Alternative would be to not draw bridge supports, but that would add other graphical weirdnesses.

**Reply by reldred** [reply]
I don't think this is strictly necessary to automatically block on greater than 1x1 - an author could always use the property to disable bridges if they don't want a pillar in the midst of their multi tile house. For instance I intend on having carparks and industrial yards, and commercial plazas and other such things spawning in one of my sets and I forsee I may end up with multi-tile variations, and these would be quite fine to have a bridge support plopped in the middle of.

**Comment by zephyris** [comment]
## More terrain variables for towns

Analogs of more _house_ variables for _towns_, which would allow houses to access parent town properties which give context like elevation, hillyness, proximity to a coastline, etc. This would allow houses to only spawn in a hilly town in the desert, or a flat town near a coast, and give a unified appearance by deciding per town rather than per house.

This is essentialy implementing property `62` of houses for towns, giving them `nearby_tile_slope`, `nearby_tile_is_water`, `nearby_tile_terrain_type`, `nearby_tile_water_class` and `nearby_tile_height`.

**Reply by zephyris** [reply]
Addressed by PR: https://github.com/OpenTTD/OpenTTD/pull/15194

**Comment by zephyris** [comment]
## Town-wide cargo acceptance/production variables

It'd be nice for NewGRFs to access more information about how much cargo is produced and transported by towns. Currently, `percent_transported_passengers` and `percent_transported_mail` are available, but this is limited (and I'm not sure how this interacts with NewGRF industries).

`number_transported_passengers` and `number_transported_mail` would be useful. Variables for `accepted_passengers` and `accepted_mail` for the previous month would be useful. Or, even better, `number_cargo` and allow querying with an arbitrary cargo ID. Houses can also produce other cargos, and accessing the quantity produced of those cargos would be useful.

**Reply by 2TallTyler** [reply]
`percent_transported_passengers` and `_mail` work fine with NewGRF industries.

House GRFs typically don't define cargo types, so I would use production effects to query transportation stats, from #11947.

**Reply by zephyris** [reply]
Wellll, maybe not publicly available House NewGRFs 😉. I've got some unpublished experiments with town houses acting like mini variable industries, which produce and consume non-passenger/mail/goods cargoes. More ways to query town production and acceptance would be nice.

**Reply by PeterN** [reply]
Town variables already exist for the amount of passengers/mail transported and produced, both this month and last month. These values don't interact with industry cargo acceptance/production at all.

Being able to specify a particular cargo type by parameter is feasible.

**Reply by zephyris** [reply]
Partially addressed by PR: https://github.com/OpenTTD/OpenTTD/pull/15195

**Comment by nielsmh** [comment]
## Presence and activity of nearby industries
Querying for the presence of either specific industries (maybe difficult to indicate which one you want) or industries meeting specific produces-cargoes/accepts-cargoes masks could let houses be selected by what kind of industry is present in or around a town. Maybe the presence of a Bank in the town would allow more office buildings to appear? Presence of a power plant in early years would give the town a more early-industrialized look? And so on.

**Comment by reldred** [comment]
I've thought long and hard about this, and I didn't think I had anything to add, but I think the only major thing I'd want beyond what some of the other folks have suggested is some additional properties/callbacks for setting whether when this building gets picked for creation is whether it can blow up everything in its path that would otherwise block it's construction. Optionally, I'd like to be able to specify whether it's allowed to knock down other 'protected' buildings. Perhaps also able to be tied to a callback so I can check the class id of a building to be knocked down so I don't accidentally clobber a 'monument'

Reason I'd want this? Large 2x2 buildings (and larger if we get them) can sometimes be difficult to spawn, and in particular if I'm trying to spawn in a new monument after a certain date, it can be hard to guarantee that it actually spawns.

**Reply by KeepinItRail** [reply]
I like this. Large 2x2 buildings better replacing smaller buildings in the city center as a city grows.

**Comment by zephyris** [comment]
## Depth and tunneling
Like adding height and a flag to allow bridging, it might be interesting to have a flag which disallows tunneling under a house. This could be extended to have a depth, and tunnels would have to be at least that deep.

This could add nice flavour, eg. disallow tunnels under a landmark, or ensure tunnels are deep under buildings with basements
It also constrains building, nice to add options to restrict building to add complexity. Features tend to make building easier.

Could be extended to objects, industries too.

**Comment by KeepinItRail** [comment]
## Population Variable

Current variable within towns ranges 0 to 65535. It is not uncommon for towns to exceed this size in which behaviour in NewGRFs starts to not work correctly.

**Comment by KeepinItRail** [comment]
## Houses on slopes only

There is a building flag HOUSE_FLAG_NOT_SLOPED which makes building in build on flat tiles. 

Can we have the a HOUSE_FLAG_SLOPED flag which does the opposite and makes a particular building only appear on slopes.

**Reply by zephyris** [reply]
This is already possible, I think(?), but takes callbacks: Use a `construction_check` callback, then decide the result based on `nearby_tile_slope` checking the with zero x, y offset.

**Reply by KeepinItRail** [reply]
Yea, possible via construction check. My thinking was whether using construction checks 'wastes' attempts at choosing a house on flat tiles, whereas a flag just cleanly removes it from consideration in the first place.
